docker/main/ngen/py_funcs.py

```diff
@@ -0,0 +1,309 @@
+
+
+def _parse_for_gather_output(parent_subparsers_container):
+    # A parser for the 'tar_and_copy' param itself, underneath the parent 'command' subparsers container
+    desc = "Using subprocesses, gather output from remote MPI hosts and collect in analogous directory on this host."
+    helper_cmd_parser = parent_subparsers_container.add_parser('gather_output', description=desc)
+    # TODO: (later) when we move to all Python, rework this to accept several individually
+    helper_cmd_parser.add_argument('mpi_hosts_str', help='Comma delimited MPI hosts string received by worker.')
+    helper_cmd_parser.add_argument('output_write_dir', type=Path, help='Directory where output was written on hosts.')
+
+
+def _subparse_move_to_directory(parent_subparser_container):
+    sub_cmd_parser = parent_subparser_container.add_parser('to_directory', description="Move to a specified directory")
+    sub_cmd_parser.add_argument("dest_dir", type=Path, help="Destination directory to which to move the output")
+
+
+def _parse_for_move_job_output(parent_subparsers_container):
+    # A parser for the 'tar_and_copy' param itself, underneath the parent 'command' subparsers container
+    desc = "Move output data files produced by a job to another location, typically to put them into a DMOD dataset."
+    helper_cmd_parser = parent_subparsers_container.add_parser('move_job_output', description=desc)
+    helper_cmd_parser.add_argument('--job_id', '--job-id', dest='job_id', help='Optionally specify job id.')
+    helper_cmd_parser.add_argument('--archive-files', dest='do_archiving', choices=["true", "false"],
+                                   type=lambda s: True if s.lower == "true" else False, default=None,
+                                   help='Force archiving before moving job output.')
+    helper_cmd_parser.add_argument('output_directory', type=Path,
+                                   help='Source directory containing output files to be placed within the dataset.')
+
+    cmd_subparsers = helper_cmd_parser.add_subparsers(dest='move_action', help="Specify the type of move action.")
+    cmd_subparsers.required = True
+    _subparse_move_to_directory(cmd_subparsers)
+
+
+def _parse_args() -> argparse.Namespace:
+    """
+    Set up and run top-level arg parsing for module.
+
+    Returns
+    -------
+    argparse.Namespace
+        The parsed arguments namespace object.
+    """
+    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter, prog='py_funcs')
+    parser.add_argument('--log-level', '-L', dest='log_level',
+                        default=os.environ.get("DEFAULT_LOG_LEVEL", "INFO").upper(),
+                        choices=["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG"], help='Optionally specify log level.')
+
+    subparsers = parser.add_subparsers(dest='command', help='Specify the Python helper function command to run.')
```

Comment:
how exclusive are these three operations supposed to be? Is this a replacement for `tar_and_copy` or a replacement for `tar_and_copy` plus other stuff?

Reply:
Initially I'd added a handful of new shell functions within _funcs.sh_, one of which was `tar_and_copy`.    I didn't want to change the entire design of the entrypoint setup too much at this stage, so I added _py_funcs.py_ and rewrote all of those new functions in Python there (removing the original additions to _funcs.sh_).  There is, of course, more in _py_funcs.py_, mostly for CLI arg parsing this file can be used directly from the existing shell scripts, but also a little bit to modularize some of the behavior.

Reply:
I'm _really_ not a fan of shoving the kitchen sink under here (rather than having `tar_and_copy.py` for instance), but if this is a case of "Perfection is the enemy of progress", I'm all for slapping a bunch of warnings around it and adding a ticket saying something along the lines of "spruce this up this way".

Reply:
Yeah, this goes back to my comment earlier about #665 and why I originally wrote things in bash.  There is no way to make this look good (at least not to all of us at once) without going waaay out of scope.  But we'll get there, and I think this approach will make that a least a little easier (given that the changes had to be done somehow).